app/services/statistics.rb

```diff
@@ -1,26 +1,30 @@
-      CountFacts.new(:count_by => :manufacturer, :title => _("Hardware"), :search => "facts.manufacturer~~VAL~"),
-      CountNumericalFactPair.new(:count_by => :memory, :title => _("Average Memory Usage")),
-      CountNumericalFactPair.new(:count_by => :swap, :title => _("Average Swap Usage")),
-      CountPuppetClasses.new(:id => :puppetclass, :title => _("Class Distribution"), :search => "class=~VAL1~")
+      CountHosts.new(:count_by => :operatingsystem, :title => ("OS Distribution"), :search =>"os_title=~VAL~", :url => build_url('operatingsystem')),
+      CountHosts.new(:count_by => :architecture, :title => _("Architecture Distribution"), :search => "facts.architecture=~VAL~", :url => build_url('architecture')),
+      CountHosts.new(:count_by => :environment, :title => _("Environment Distribution"), :search => "environment=~VAL~", :url => build_url('environment')),
+      CountHosts.new(:count_by => :hostgroup, :title => _("Host Group Distribution"), :search => "hostgroup=~VAL~", :url => build_url('hostgroup')),
+      CountHosts.new(:count_by => :compute_resource, :title => _("Compute Resource Distribution"), :search => "compute_resource=~VAL~", :url => build_url('compute_resource')),
+      CountFacts.new(:count_by => :processorcount, :unit => Nn_('%s core', '%s cores'), :title => _("Number of CPUs"), :search => "facts.processorcount=~VAL1~", :url => build_url('processorcount')),
+      CountFacts.new(:count_by => :manufacturer, :title => _("Hardware"), :search => "facts.manufacturer~~VAL~", :url => build_url('manufacturer')),
+      CountNumericalFactPair.new(:count_by => :memory, :title => _("Average Memory Usage"), :url => build_url('memory')),
+      CountNumericalFactPair.new(:count_by => :swap, :title => _("Average Swap Usage"), :url => build_url('swap')),
+      CountPuppetClasses.new(:id => :puppetclass, :title => _("Class Distribution"), :search => "class=~VAL1~", :url => build_url('puppetclass'))
     ]
 
     if SETTINGS[:locations_enabled]
-      charts << CountHosts.new(:count_by => :location, :title => _("Location Distribution"), :search => "location=~VAL~")
+      charts << CountHosts.new(:count_by => :location, :title => _("Location Distribution"), :search => "location=~VAL~", :url => build_url('location'))
     end
 
     if SETTINGS[:organizations_enabled]
-      charts << CountHosts.new(:count_by => :organization, :title => _("Organization Distribution"), :search => "organization=~VAL~")
+      charts << CountHosts.new(:count_by => :organization, :title => _("Organization Distribution"), :search => "organization=~VAL~", :url => build_url('organization'))
     end
 
     charts
   end
+
+  def self.build_url(id)
+    Rails.application.routes.url_helpers.url_for :controller => 'statistics', :action => 'show', :id => id, :only_path => true, :organization_id => Organization.current.try(:id), :location_id => Location.current.try(:id)
```

Comment:
Nice find on the url_helpers that we can use here without the view, but if we aren't pushing it up the stack to the controller, we can leave it at the `Statistics::Base` level. My main concern was with building the url ourselves which is why I suggested pushing it up to where we have the url helpers. 
I suggest setting it in the `Statistics::Base` initializer unless a url is explicitly specified to reduce the extra parameter passing the same id we already have, so something like:
```ruby
class Base
...
  def initilize(options = {})
...
    @url = options[:url] || build_url
...
  end
...
  def build_url
    Rails.application.routes.url_helpers.url_for :controller => 'statistics', :action => 'show', :id => @id, :only_path => true, :organization_id => Organization.current.try(:id), :location_id => Location.current.try(:id)
  end
...
```

This way we preserve the ability to pass a custom url if needed (that we don't do now), but default to always including the taxonomies in the default url. (and also reduce the diff even more in case backporting is needed)

Reply:
all right, I think it's nice to be able to render a chart for other than current org/loc so I don't load the current context inside the statistic but rather pass it to the initializer, but the rest of the logic is now back